### PR TITLE
Conditional event system

### PR DESCRIPTION
**Pull request by Frickerson**

### CODE REVIEW

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include "Abilities/GASExtGameplayAbility.h"
+
+#include <CoreMinimal.h>
+
+#include "GBFConditionalEvent.generated.h"
+
+class UGBFConditionalTrigger;
+
+UCLASS( HideDropdown )
+class GAMEBASEFRAMEWORK_API UGBFConditionalEvent : public UGASExtGameplayAbility
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: Suffix with `Ability`

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    void EndAbility( const FGameplayAbilitySpecHandle handle,
+        const FGameplayAbilityActorInfo * actor_info,
+        const FGameplayAbilityActivationInfo activation_info,
+        bool replicate_end_ability,
+        bool was_cancelled ) override;
+
+protected:
+    UFUNCTION( BlueprintImplementableEvent, Category = "Conditional Events" )
+    void ApplyOutcomes();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: what about `ExecuteOutcomes` ?

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+private:
+    UFUNCTION()
+    void OnTriggerTriggered();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: `OnTriggersTriggered`

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "GBFConditionalEventSubsystem.generated.h"
+
+class UGBFConditionalEvent;
+class UGBFConditionalEventGroupData;
+class UAbilitySystemComponent;
+class AGameStateBase;
+
+UCLASS()
+class GAMEBASEFRAMEWORK_API UGBFConditionalEventSubsystem : public UWorldSubsystem
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h:15]: Implement `DoesSupportWorldType` (check the game phase subsystem)

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#include "Gameplay/GBFTriggerBox.h"
+
+#include <CoreMinimal.h>
+#include <UObject/Object.h>
+
+#include "GBFConditionalTrigger.generated.h"
+
+DECLARE_DYNAMIC_MULTICAST_DELEGATE( FGBFOnTriggeredDelegate );
+
+UCLASS( NotBlueprintable, EditInlineNew, HideDropdown )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Abstract

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+class GAMEBASEFRAMEWORK_API UGBFConditionalTrigger : public UObject
+{
+    GENERATED_BODY()
+
+public:
+    virtual void Activate();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: PURE_VIRTUAL (both)

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    FGBFOnTriggeredDelegate OnTriggeredDelegate;
+};
+
+FORCEINLINE FGBFOnTriggeredDelegate & UGBFConditionalTrigger::GetOnTriggeredDelegate()
+{
+    return OnTriggeredDelegate;
+}
+
+UCLASS()
+class GAMEBASEFRAMEWORK_API UGBFTriggerBoxTrigger : public UGBFConditionalTrigger
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Create a `Triggers` subfolder and put it there (you can leave the base class here)

Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: final

File: Source/GameBaseFramework/Public/Gameplay/GBFTriggerBox.h [hunk 8]
```diff
@@ -23,6 +23,8 @@ class GAMEBASEFRAMEWORK_API AGBFTriggerBox : public AActor
     UFUNCTION( NetMulticast, reliable, BlueprintCallable )
     void Multicast_DeactivateTrigger();
 
+    UGBFTriggerManagerComponent * GetTriggerManagerComponent() const;
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Gameplay/GBFTriggerBox.h]: inline functions are declared first

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
+#include "Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h"
+
+#include "GameFramework/GBFGameState.h"
+#include "Gameplay/ConditionalEvents/GBFConditionalEvent.h"
+#include "Gameplay/ConditionalEvents/GBFConditionalEventGroupData.h"
+
+#include <AbilitySystemComponent.h>
+
+bool UGBFConditionalEventSubsystem::ShouldCreateSubsystem( UObject * outer ) const
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: On Lyra, the phase subsystem does not implement that function. The system would be created everywhere. But you need to add `BlueprintAuthorityOnly` on the BP functions

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    return true;
+}
+
+void UGBFConditionalEventSubsystem::Initialize( FSubsystemCollectionBase & collection )
+{
+    Super::Initialize( collection );
+
+    if ( auto * world = GetWorld() )
+    {
+        world->GameStateSetEvent.AddUObject( this, &ThisClass::OnGameStateSet );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: with game features being enabled / disabled at runtime, you can't cache the game state ASC like that.
You will need to get it like that everytime you need it : `auto * game_state_asc = world->GetGameState()->FindComponentByClass< UGASExtAbilitySystemComponent >();`

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    if ( !ensureMsgf( conditional_event != nullptr, TEXT( "%s: Conditional event is not valid!" ), TEXT( __FUNCTION__ ) ) )
+    {
+        return;
+    }
+
+    if ( const auto * spec = GameStateASC->FindAbilitySpecFromClass( conditional_event ) )
+    {
+        GameStateASC->CancelAbilityHandle( spec->Handle );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: you need to remove the Handle from the array

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+void UGBFConditionalEventSubsystem::DeactivateAll()
+{
+    if ( !ensureMsgf( GameStateASC != nullptr, TEXT( "%s: Game State ASC not valid!" ), TEXT( __FUNCTION__ ) ) )
+    {
+        return;
+    }
+
+    for ( auto & event_handle : ActivatedEventHandles )
+    {
+        GameStateASC->CancelAbilityHandle( event_handle );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: clear the array of handles

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    }
+}
+
+void UGBFConditionalEventSubsystem::OnGameStateSet( AGameStateBase * game_state_base )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: should be removed

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+void UGBFTriggerBoxTrigger::Activate()
+{
+    if ( !ensureMsgf( !TriggerBoxSoftObject.IsNull(), TEXT( "%s: TriggerBox reference is not valid!" ), TEXT( __FUNCTION__ ) ) )
+    {
+        return;
+    }
+
+    auto * trigger_box = TriggerBoxSoftObject.Get();
+    trigger_box->Multicast_ActivateTrigger( true );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp]: I disagree with this. We will never find out why a disabled trigger suddenly gets activated. If the trigger is not active, so be it, the trigger won't execute

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+void UGBFTriggerBoxTrigger::Deactivate()
+{
+    if ( !ensureMsgf( !TriggerBoxSoftObject.IsNull(), TEXT( "%s: TriggerBox reference is not valid!" ), TEXT( __FUNCTION__ ) ) )
+    {
+        return;
+    }
+
+    auto * trigger_box = TriggerBoxSoftObject.Get();
+    trigger_box->GetTriggerManagerComponent()->OnTriggerBoxActivated().RemoveDynamic( this, &ThisClass::OnTriggerBoxActivated );
+    trigger_box->Multicast_DeactivateTrigger();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp]: same

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h (hunk 1 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: Done.

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h (hunk 2 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: Indeed a bit better.

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h (hunk 3 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEvent.h]: Done.

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h (hunk 4 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.h:15]: Done

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h (hunk 7 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Done.

Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Done.

File: Source/GameBaseFramework/Public/Gameplay/GBFTriggerBox.h (hunk 8 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/GBFTriggerBox.h]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 9 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 10 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 11 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 12 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 13 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp (hunk 14 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp (hunk 15 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalTrigger.cpp]: Done.

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h (hunk 6 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Done.

File: Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h (hunk 5 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Gameplay/ConditionalEvents/GBFConditionalTrigger.h]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+    FGameplayAbilitySpec spec( conditional_event );
+
+    auto handle = game_state_asc->GiveAbilityAndActivateOnce( spec );
+    ActivatedEventHandles.Emplace( handle );
+}
+
+void UGBFConditionalEventSubsystem::ActivateEventGroup( const UGBFConditionalEventGroupData * conditional_event_group_data )
+{
+    const auto * game_state_asc = GetGameStateASC();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: you don't need lines from L36 to L42

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        auto & handle = spec->Handle;
+        game_state_asc->CancelAbilityHandle( handle );
+        ActivatedEventHandles.Remove( handle );
+    }
+}
+
+void UGBFConditionalEventSubsystem::DeactivateEventGroup( const UGBFConditionalEventGroupData * conditional_event_group_data )
+{
+    const auto * game_state_asc = GetGameStateASC();
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: don't need ASC here

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 17 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

File: Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp (hunk 16 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Private/Gameplay/ConditionalEvents/GBFConditionalEventSubsystem.cpp]: Done.

**Review by TheEmidee — APPROVED** [review]

**Pull request merged**